src/sources/mongodb_metrics/mod.rs

```diff
@@ -0,0 +1,1065 @@
+            serde_json::to_string(&build_info).unwrap()
+        );
+
+        Ok(Arc::new(Self {
+            client,
+            endpoint,
+            namespace: namespace.to_owned(),
+            tags,
+        }))
+    }
+
+    /// Remove `username` and `password` from URL endpoint.
+    fn sanitize_endpoint(endpoint: &str) -> Result<String, BuildError> {
+        let mut url = Url::parse(endpoint).context(ParseEndpointError)?;
+        url.set_username("")
+            .map_err(|_| BuildError::SanitizeEndpointError {
+                error: "failed to remove username from MongoDB endpoint",
+            })?;
+        url.set_password(None)
+            .map_err(|_| BuildError::SanitizeEndpointError {
+                error: "failed to remove password from MongoDB endpoint",
+            })?;
+        Ok(url.to_string())
+    }
+
+    /// Finding node type for client with `isMaster` command.
+    async fn get_node_type(client: &Client) -> Result<NodeType, BuildError> {
+        let doc = client
+            .database("admin")
+            .run_command(doc! { "isMaster": 1 }, None)
+            .await
+            .context(CommandIsMasterMongoError)?;
+        let msg: CommandIsMaster = from_document(doc).context(CommandIsMasterParseError)?;
+
+        Ok(if msg.set_name.is_some() || msg.hosts.is_some() {
+            NodeType::Replset
+        } else if msg.msg.map(|msg| msg == "isdbgrid").unwrap_or(false) {
+            // Contains the value isdbgrid when isMaster returns from a mongos instance.
+            // https://docs.mongodb.com/manual/reference/command/isMaster/#isMaster.msg
+            // https://docs.mongodb.com/manual/core/sharded-cluster-query-router/#confirm-connection-to-mongos-instances
+            NodeType::Mongos
+        } else {
+            NodeType::Mongod
+        })
+    }
+
+    async fn get_build_info(client: &Client) -> Result<CommandBuildInfo, BuildError> {
+        let doc = client
+            .database("admin")
+            .run_command(doc! { "buildInfo": 1 }, None)
+            .await
+            .context(CommandBuildInfoMongoError)?;
+        from_document(doc).context(CommandBuildInfoParseError)
+    }
+
+    fn encode_namespace(&self, name: &str) -> String {
+        match self.namespace.as_str() {
+            "" => name.to_string(),
+            _ => format!("{}_{}", self.namespace, name),
+        }
+    }
+
+    fn create_metric(
+        &self,
+        name: &str,
+        value: MetricValue,
+        tags: BTreeMap<String, String>,
+    ) -> Metric {
+        Metric {
+            name: self.encode_namespace(name),
+            timestamp: Some(Utc::now()),
+            tags: Some(tags),
+            kind: MetricKind::Absolute,
+            value,
+        }
+    }
+
+    async fn collect(self: Arc<Self>) -> stream::BoxStream<'static, Metric> {
+        let (up_value, metrics) = match Arc::clone(&self).collect_server_status().await {
+            Ok(metrics) => (1.0, metrics),
```

Comment:
This kind of puzzled me a bit, but I think I understand now. A comment explaining why you return the `1.0` or `0` might help future readers.